docs/gavo.1
.\" Man page generated from reStructuredText.
.
.TH DACHS 1 "2017-01-18" "1.0" ""
.SH NAME
dachs \- data publishing infrastructure for the Virtual Observatory (VO)
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
dachs [global\-options] <subcommand> [options] function\-argument ...
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
dachs  provides  support for data ingestion and publishing, for metadata
handling, and for a variety of VO protocols and standards, e.g. the
Table Access Protocol (TAP) or the Simple Cone Search (SCS).
.sp
There are numerous sub\-commands covering the various tasks (importing,
controlling the server, running tests, etc).
.sp
Subcommand names can be abbreviated to the shortest unique prefix.
.sp
A centeral concept of DaCHS is the Resource Descriptor (RD), and XML
description of a data collection including metadata, ingestion rules,
service definitions, and regression tests.  They are usually referenced
through their RD ids, which are the relative paths from DaCHS\(aq inputs
directory to the file containing the RD, with the conventional extension
\fB\&.rd\fP stripped.  For instance, in a default install, the file
\fB/var/gavo/inputs/myrsc/q.rd\fP would have \fBmyrsc/q\fP as RD id.
.sp
Most commands dealing with RD ids will also pick up RDs if referenced by
path; in the example above, if you are in \fB/var/gavo/inputs/myrsc\fP,
you could also reference the RD as either \fBq\fP or \fBq.rd\fP\&.
.sp
Several commands take references to RD elements (table definitions, exec
items, direct grammar, etc).  These consist of an RD id as just
discussed, a hash mark, and the XML id of the target element.  Tables
have an id automatically, for other elements you may have to add an
artificial id.
.SH GLOBAL OPTIONS
.sp
Global options are given before the subcommand name.
.INDENT 0.0
.TP
.B \-\-debug
produce debug info as appropriate
.TP
.B \-\-enable\-pdb
run pdb on all errors
.TP
.B \-h\fP,\fB  \-\-help
show this help message and exit
.TP
.B \-\-hints
if there are hints on an error, display them
.TP
.BI \-\-profile\-to\fB= PROFILEPATH
enable profiling and write a profile to PROFILEPATH
.TP
.B \-\-suppress\-log
suppress logging of exceptions to the dachs\-specific log files
.UNINDENT
.SH THE ADMIN SUBCOMMAND
.sp
Synopsis:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dachs admin  [\-h] subsubfunction [subfunction\-arguments ...]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This is a somewhat random collection of commands related to
administering a data center.  In particular, this is where you create
and edit accounts.
.sp
subsubcommands can be abbreviated as long as the abbreviation is
unique.  For instance \fBgavo adm xsd\fP will do an XSD validation.
.sp
For more information on the subsubfunctions, pass a \fB\-h\fP flag.
.SS Subsubcommands
.INDENT 0.0
.IP \(bu 2
\fBaddtogroup user group\fP \-\- adds a user to a group
.IP \(bu 2
\fBadduser user password [remarks]\fP \-\- add  a  user/password pair to
the DaCHS user table.  Note that as of DaCHS 1.0 the password is stored
in clear text and also transmitted in clear text since DaCHS only
supports HTTP basic authentication.  Do not use valuable passwords
here
.IP \(bu 2
\fBdelfromgroup user group\fP \-\- remove a user from a group
.IP \(bu 2
\fBdeluser user\fP \-\- remove a DaCHS user from the user table.
.IP \(bu 2
\fBlistusers\fP \-\- dump the user table
.IP \(bu 2
\fBchangeuser user password [remarks]\fP \-\- change remarks and/or
password for a DC user.  See adduser for details.
.IP \(bu 2
\fBcleantap\fP \-\- remove expired Universal Worker Service (UWS) jobs
.IP \(bu 2
\fBtapabort jobId helpMsg\fP \-\- manually abort a TAP job and return
helpMsg to the requesting user as error message.
.IP \(bu 2
\fBupdateTAPSchema\fP \-\- Update the TAP_SCHEMA metadata for all RDs
mentioned in TAP_SCHEMA.
.IP \(bu 2
\fBdeclaredel\fP \-\- declare  an identifier as deleted (for when you\(aqve
removed the RD but the identifier still floats  on some registries.
.IP \(bu 2
\fBindexStatements tableRef\fP \-\- show the statements to create the
indices on a table.  The tableRef has the format RD\-id#table\-id; it
is \fInot\fP a database table reference.
.IP \(bu 2
\fBdumpDF path\fP \-\- Dumps the source of a file included with the DaCHS
distribution.  The argument is a package resource path.
(like /inputs/__system__/scs.rd); for system RDs, the special
//rd\-id syntax is supported.
.IP \(bu 2
\fBexecute execId\fP \-\- Execute the contents of an RD execute element. You
must give that element an explicit id in order to make this work; then
execId is rd\-id#exec\-id
.IP \(bu 2
\fBxsdValidate path\fP \-\- Validate a file against built\-in VO schemas
and with built\-in schema validator.
.IP \(bu 2
\fBsuggestucds tableId\fP \-\- Make suggestions for UCDs of the columns
of the referenced table (rd\-id#table\-id format) not having one.
This is based on their descriptions and uses a GAVO web service.
.UNINDENT
.SH THE CONFIG SUBCOMMAND
.sp
Synopsis:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dachs config [section\-name] config\-key
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This outputs values of DaCHS\(aq configuration to stdout.  section\-name
defaults to general.  This is most commonly used to make external
components aware of DaCHS\(aq file locations, e.g., through
\fBinputs_dir=$(dachs config inputsDir)\fP\&.
.sp
See the operator\(aqs guide for a documentation on DaCHS\(aq configuration
options.
.SH THE DROP SUBCOMMAND
.sp
Synopsis:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
drop [\-h] [\-s] [\-\-all] rd\-id [dd\-id ...]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This is the reverse of import: Tables created by a \fBdachs imp\fP with
identical arguments are being torn down by \fBdachs drop\fP\&.  This will
not work reliably if the RD has been modified between the imp and the
drop, in particular if the RD has been deleted.  In such situations, you
can use the \fB\-f\fP flag, which unconditionally tears down everything
DaCHS has recorded as coming from the referenced RD.
.SS Arguments
.INDENT 0.0
.TP
.B rd\-id
RD path or ID to drop
.TP
.B dd\-id
optional  data descriptor (DD) ID(s) if you do not
want to drop the entire RD; note that  no  service
publications will be undone if you give DD IDs
.UNINDENT
.SS Options
.INDENT 0.0
.TP
.B \-\-all
drop  all  DDs  in  the RD, not only the auto ones
(overrides manual selection)
.TP
.B \-s\fP,\fB  \-\-system
drop tables even if they are system tables
.UNINDENT
.SH THE IMPORT SUBCOMMAND
.sp
Synopsis:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
import [option] rd\-name [data\-id]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This subcommand is used to ingest data described by an RD.  For special
applications, ingestion can be restricted to specific data items within
an RD.
.SS Options
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
show this help message and exit
.TP
.B \-n\fP,\fB  \-\-updateRows
Deprecated.  Use updating data items instead.
.TP
.B \-d\fP,\fB  \-\-dumpRows
Dump raw rows as they are emitted by the grammar.
.TP
.B \-D\fP,\fB  \-\-dumpIngestees
Dump processed rows as emitted by the row makers.
.TP
.B \-R\fP,\fB  \-\-redoIndex
Drop indices before updating a table and recreate them
when done
.TP
.B \-m\fP,\fB  \-\-meta\-only
just update table meta (privileges, column
descriptions,...).
.TP
.B \-I\fP,\fB  \-\-meta\-and\-index
do not import, but update table meta (privileges,
column descriptions,...) and recreate the indices
.TP
.B \-u\fP,\fB  \-\-update
update mode \-\- don\(aqt drop tables before writing.
.TP
.B \-s\fP,\fB  \-\-system
(re\-)create system tables, too
.TP
.B \-v\fP,\fB  \-\-verbose
talk a lot while working
.TP
.B \-r\fP,\fB  \-\-reckless
Do not validate rows before ingestion
.TP
.BI \-M \ MAX\fP,\fB \ \-\-stop\-after\fB= MAX
Stop after having parsed MAX rows
.TP
.BI \-b \ N\fP,\fB \ \-\-batch\-size\fB= N
deliver N rows at a time to the database.
.TP
.B \-c\fP,\fB  \-\-continue\-bad
do not bail out after an error, just skip the current
source and continue with the next one.
.TP
.B \-L\fP,\fB  \-\-commit\-after\-meta
commit the importing transaction after updating the
meta tables.  Use this when loading large (hence \-L)
data sets to avoid keeping a lock on the meta tables
for the duration of the input, i.e., potentially days.
The price is that users will see empty tables during
the import.
.UNINDENT
.SH THE INFO SUBCOMMAND
.sp
Synopsis:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
info [\-h] table\-id
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This displays column statistics about the table referred to in the
argument (which must be a fully qualified table name resolvable by the
database system).
.SS Argument
.INDENT 0.0
.TP
.B table\-id
table ID (of the form rdId#tableId)
.UNINDENT
.SH THE INIT SUBCOMMAND
.sp
Synopsis:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
init [\-h] [\-d DSN] [\-\-nodb]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This initialises DaCHS\(aq file system and database environment.  Calling
\fBdachs init\fP on an existing site should not damage anything.  It
might, however, fix things if, for instance, permissions on some
directories went funny.
.SS Options
.INDENT 0.0
.TP
.BI \-d \ <DSN>\fP,\fB \ \-\-dsn \ <DSN>
data source name (DSN) to use to  connect  to  the
future DaCHS database; the DSN must let DaCHS con‐
nect to the database as an administrator;  dbname,
host,  and  port  get  copied  to  the profile, if
given; if you followed the  installation  instruc‐
tions, you do not need this option
.TP
.B \-\-nodb

Changes
Note: Of course, in each version, many minor improvements and bug fixes 
go in.  To see the full history, use  

svn log http://svn.ari.uni-heidelberg.de/svn/gavo/python/trunk/ | less

to see the gory details.  If anything not mentioned here breaks something
for you, we'd consider it a severe bug, though.

Version 0.9.8 (2016-12-08)

	* SODA/Datalink implementation is now considered stable.  Go forth and
	  make fabulous links and cutout services.  Check the overhauled datalink 
	  section in the reference documentation if you have a couple of minutes:
		http://docs.g-vo.org/DaCHS/ref.html#datalink-and-soda

	* TAP_SCHEMA columns now has new column_index column.  Its addition should
	  be automatic as long as your TAP-published RDs are in order.

	* EPN-TAP 2.0 changed some column and parameter names, and the mixin
	  followed.  If you've already used EPN-TAP 2.0, you will have to re-import
	  your data and slightly adapt your RD (e.g., c1_min -> c1min, 
	  t_exp_min -> time_exp_min etc).

	* DaCHS now emits warnings for columns whose names are SQL or ADQL 
		reserved names (this includes popular items like area, size, and distance).
		Even if things have worked so far, please change these names.  We 
		guarantee you'll regret it if you don't.

	* Full support for VOResource 1.1 relationships 
		(http://docs.g-vo.org/vocab-test/relationship_type).  Please try them
		if you have appropriate data.


Version 0.9.7 (2016-08-18)

	* Behaviour change: pythonCores no longer run in a file system sandbox 
		(can't do that thread-safely).

	* Support for SIAP version 2 (see tutorial for details)

	* The default VOTable version delivered is now VOTable 1.3 (actually,
	  it's been 1.3 for a while now, we just didn't declare it)

	* support for Obscore 1.1 (update should be automatic)

	* support for EPN-TAP 2.0 (manually change to the #table-2.0 mixin)

	* Obscore support is now at version 1.1

	* DaCHS now supports CORS for web-based UIs

	* DaCHS's ADQL cross matches should now have "good" query plans 
	  regardless or argument order (but are half as  fast as in previous.
	  "good" cases).

	* SCS services publishing TAP-readable tables now get an auxiliary
	  TAP publication automatically (you'll need to re-gavo pub them,
	  though, if you want the Registry to notice quickly).

	* Removed bbox-based SIAP because DaCHS has been requiring pgsphere 
	  for quite a while now, and hence there's not need for a lousy fallback.

	* gavo admin xsdVal and XSD-validating regression tests now work
	  even with package installations; so: you got your own schema mirror now!

	* New gavo admin updateTAPSchema command for syncing TAP_SCHEMA with
	  the RDs.

	* New built-in preview-managing scheme for large data collections.

	* As usual, many small bug fixes and new features.


Version 0.9.6 (2016-03-16)

  * Several new indices are now created by default; you probably want them,
    too.  Run "gavo imp -I //products". Similarly, gavo imp -I <rd-id> on all
    RDs that have SSA tables with datalink services (if you don't know what
    that is, you don't have it)

  * New soda RD containing (almost) all to build a WD-compliant cube
    cutout service (and a bit more).  You can update any existing
    services you have by changing "datalink" to "soda" in the referenced
    procDefs.

  * API change: base.makeMetaItem and makeMetaValue were a pain.  They
		have simply gone in the expectation that nobody has actually used them.
		Protest if we're wrong.  Also there is no type attribute on meta
		any more.  Also, no meta/@type any more in RDs.  I hope you've not
		really used *that*.

	* API change: When CondDescs yielded more than one fragement,
	  the fragments were combined with AND so far.  Now, it's OR by
	  default (you can restore the previoius behaviour using condDesc/@joiner)

	* Some changes in the way the selection and evaluation of sort keys work
	  in particular in web interfaces.  If you've been playing games with this,
	  check if what you did still works.

	* Support for DALI 1.1 xtype="interval" parameters.

	* New recommended way to specify author lists; see
	  http://docs.g-vo.org/DaCHS/tutorial.html#authors

	* As usual, many small bug fixes and new features.


Version 0.9.5 (2015-10-07)

	* **RD breakage**: The old DaCHS implemenation got the meaning of
		spectralSI and fluxSI completely wrong.  You will fairly certainly have
		to change RDs doing SSA.  Any occurrence of spectralSI will have to become
		spectralUnit, and fluxSI fluxUnit (run gavo val ALL to catch RDs where
		that's necessary).  While you could add correct spectralSI and fluxSI
		values (which would look like "10e-10 L" or so), the DaCHS authors
		currently consider that a waste of time.

	* Possible service breakage: We're now using VOTable tabledata parsers
	  to parse the values of PARAMs; the same code is used to parse HTTP
	  parameters.  If you define custom InputKeys and got metadata slightly

gavo/resources/web/js/gavo.src.js
// javascript support code for the GAVO data center


////////////////////////////// possibly generic functions

function decodeGetPars(queryString) {
// an incredibly crappy approach to getting whatever was in the query string
// into javascript.
	var pars = new Object();
	var pairs = queryString.slice(1).split("&");
	for (var ind in pairs) {
		var pair = pairs[ind].split("=");
		var key = 'arg'+unescape(pair[0]).replace("+", " ");
		var value = unescape(pair[1]).replace("+", " ");
		if (pars[key]==undefined) {
			pars[key] = new Array();
		}
		pars[key].push(value);
	}
	return pars;
}


function isIn(item, arr) {
// does a linear search through arr to see if item is in there
// (can't we use assoc. arrs for that?)
	for (var ind in arr) {
		if (arr[ind]==item) {
			return true;
		}
	}
	return false;
}


function getJD(date) {
	return date/86400000+2440587.5;
}


function getJYear(date) {
	return (getJD(date)-2451545)/365.25+2000;
}


///////////// Micro templating.  See develNotes
function htmlEscape(str) {
	return String(str).replace(/&/g, '&amp;').replace(/"/g, '&quot;')
		.replace(/'/g, '&apos;').replace(/</g, '&lt;')
		.replace(/>/g, '&gt;');
}

(function () {
	var _tmplCache = {};
	this.renderTemplate = function (templateId, data) {
		var err = "";
		var func = _tmplCache[templateId];
		if (!func) {
			str = document.getElementById(templateId).innerHTML;
			var strFunc =
				"var p=[],print=function(){p.push.apply(p,arguments);};"
				+ "with(obj){p.push('"
				+ str.replace(/[\r\t\n]/g, " ")
				.split("'").join("\\'")
				.split("\t").join("'")
				.replace(/\$([a-zA-Z_]+)/g, "',htmlEscape($1),'")
				.replace(/\$!([a-zA-Z_]+)/g, "',$1,'")
				+ "');}return $.trim(p.join(''));";
				func = new Function("obj", strFunc);
				_tmplCache[str] = func;
		}
		return func(data);
	}
})()




///////////// Code for generating GET-URLs for forms

function getEnclosingForm(element) {
// returns the form element immediately enclosing element.
	if (element.nodeName=="FORM") {
		return element;
	}
	return getEnclosingForm(element.parentNode);
}

function getSelectedEntries(selectElement) {
// returns an array of all selected entries from a select element 
// in url encoded form
	var result = new Array();
	var i;

	for (i=0; i<selectElement.length; i++) {
		if (selectElement.options[i].selected) {
			result.push(selectElement.name+"="+encodeURIComponent(
				selectElement.options[i].value))
		}
	}
	return result;
}

function makeQueryItem(element) {
// returns an url-encoded query tag item out of a form element
	var val=null;

	switch (element.nodeName) {
		case "INPUT":
		case "TEXTAREA":
			if (element.type=="radio" || element.type=="checkbox") {
				if (element.checked) {
					val = element.name+"="+encodeURIComponent(element.value);
				}
			} else if (element.name && element.value) {
				val = element.name+"="+encodeURIComponent(element.value);
			}
			break;
		case "SELECT":
			return getSelectedEntries(element).join("&");
			break;
		case "BUTTON":  // no state here
			break;
		default:
			alert("No handler for "+element.nodeName);
	}
	if (val) {
		return val;
	} else {
		return element.NodeName;
	}
}

function getFormQuery(form, ignoreNames) {
	// returns a link to the result sending the HTML form form would
	// yield.
	var fragments = new Array();
	var i;

	items = form.elements;
	for (i=0; i<items.length; i++) {
		var fragment = makeQueryItem(items[i]);
		if (fragment && ignoreNames[items[i].name]==undefined) {
			fragments.push(fragment);
		} 
	}
	return form.getAttribute("action")+"?"+fragments.join("&");
}


function makeResultLink(form) {
	return getFormQuery(form, []);
}


function makeBookmarkLink(form) {
	return getFormQuery(form, {'__nevow_form__': 1});
}


///////////// Functions for the sidebar

var expandedMetaWidth = '200px';

function markFold(title, markstr) {
	return title.slice(0, title.length-2)+markstr;
}

function toggleCollapsedMeta(el) {
	var contentBox = el.parentNode.nextSibling;
	if (contentBox.style.visibility=='visible') {
		el.childNodes[0].data = markFold(el.childNodes[0].data, ">>");
		collapseMeta(contentBox);
	} else {
		el.childNodes[0].data = markFold(el.childNodes[0].data, "<<");
		expandMeta(contentBox);
	}
}

function collapseMeta(box) {
	box.style.visibility = 'hidden';
	box.style.height = '0px';
	box.style.maxHeight = '0px';
	var parent = box.parentNode;
	parent.style.border = '0px none #707ca0';
	parent.style.padding = '1px';
	parent.style.background = 'none';
}

function expandMeta(box) {
	box.style.visibility = 'visible';
	box.style.height = 'auto';
	box.style.maxHeight = '200px';
	box.style.width = expandedMetaWidth;
	var parent = box.parentNode;
	parent.style.border = '1px solid #707ca0';
	parent.style.padding = '3px';
	parent.style.background = '#ffffff';
	parent.style.width = expandedMetaWidth;
}

function followEmbeddedLink(parent) {
// forwards the browser to the href of the first child that has one.
// this is for the button-type decoration around links.
	for (index in parent.childNodes) {
		child = parent.childNodes[index];
		if (child.href) {
			window.location = child.href;
			break;
		}
	}
}

///////////// Functions dealing with the output format widget
// This incredibly verbose crap hides and shows widgets selecting aspects
// of the output format.  Basically, you have widgets in output_bussedElements
// that get notified when the output format changes and then (un)attach 
// themselves to a container.
//
// To use this, you need:
//  * a block element with id "genForm-_OUTPUT" in which the subwidgets are 
//    displayed
//  * a form element calling output_broadcast(this.value) on a change
//
//  In the DC, the static QueryMeta method getOutputWidget worries about this.


function output_BussedElement(domNode, id, visibleFor) {
	// An element that can be passed to output_broadcast and does something
	// in response.
	domNode.id = id;
	for (var ind in visibleFor) {
		domNode["visibleFor_"+visibleFor[ind]] = true;
	}
	return domNode;
}


function output_verbSelector(pars) {
	// returns a BussedElement for the selector for output verbosity
	var verbosities = new Array("H", "1", "2", "3");
	var root = document.createElement("span")
	var sel = document.createElement("select");
	var curSetting;

	root["class"] = "op_widget";
	root.appendChild(document.createTextNode(" output verbosity "));
	sel.name = "_VERB";
	if (pars['arg'+sel.name]!=undefined) {
		curSetting = pars['arg'+sel.name][0];
	} else {
		curSetting = "H";
	}
	for (verbInd in verbosities) {
		var el = sel.appendChild(document.createElement("option"));
		var verb = verbosities[verbInd];
		el.appendChild(document.createTextNode(verb));
		if (verb==curSetting) {
			el.selected = "selected";
		}
	}
	root.appendChild(sel);
	return output_BussedElement(root, "op_verb", ["VOTable", "FITS",
		"TSV"]);
}


function output_tdEncSelector(pars) {
	// returns a BussedElement to select VOTable encoding
	var root = document.createElement("span");
	var box = document.createElement("input");
	var curSetting;

	root["class"] = "op_widget";
	box.name = "_TDENC";
	if (pars['arg'+box.name]!=undefined) {
		curSetting = pars['arg'+box.name][0];
	} else {
		curSetting = "false";
	}
	box.type = "checkbox";
	box.style.width = "auto";
	root.appendChild(box);
	root.appendChild(document.createTextNode(" human-readable "));
	if (curSetting=="true") {
		box.checked = "checked";
	}
	return output_BussedElement(root, "op_tdenc", ["VOTable"]);
}


function output_setFormat(format) {
	var opts=document.getElementById("genForm-_FORMAT").options;
	for (var optInd=0; optInd<opts.length; optInd++) {
		if (opts[optInd].value==format) {
			opts[optInd].selected = true;
		} else {
			opts[optInd].selected = false;
		}
	}
}

var output_bussedElements = new Array();


function output_init() {
	var pars = decodeGetPars(location.search);
	var format = pars["arg_FORMAT"];

	if (!document.getElementById("genForm-_OUTPUT")) { // no form on page
		return;
	}
	output_bussedElements.push(output_verbSelector(pars));
	output_bussedElements.push(output_tdEncSelector(pars));
	output_bussedElements.push(output_itemSelector(pars));
	if (format==undefined) {
		format = "HTML";
	}
	output_broadcast(format);
	outputInited = true;
}

function output_broadcast(newFormat) {
	var visibleForThis = "visibleFor_"+newFormat;

	output_setFormat(newFormat);
	for (var ind in output_bussedElements) {
		var el=output_bussedElements[ind];
		if (el[visibleForThis]) {
			output_show(el);
		} else {
			output_hide(el);
		}
	}
}


$(document).ready(
	function() {
		output_init();
	});



///////////////// jquery-dependent code (TODO: move the rest of the
/// stuff from plain js to jquery, too)


function makeTabCallback(handlers) {
// see develNotes
	return function (ev) {
		$("#tabset_tabs > li").removeClass("selected");
		var curTab = $(ev.currentTarget);
		curTab.addClass("selected");
		var mainId = curTab.find("a").attr("name");
		handlers[mainId]($('#mainbox'));
	}
}


function popupInnerWindow(content, parent, onClose) {
// puts content into a thing that looks like a window and can be
// dragged and resized.  Parent is the element the container
// gets added to within the tree.
// The function returns a closer function destroying the subwindow.
	var prevParent = content.parent()
	prevParent.css({"position": "relative"});
	content.detach();
	content.css("visibility", "visible");

	var container = $("<div class='innerWin'/>");
	var titlebar = $(
		"<p class='innerTitle'><span class='closer'>x&nbsp;</span></p>");
	container.append(titlebar);
	container.append(content);
	content.addClass("innerBody");
	content.show();

	container.draggable({cancel: ".innerBody"});
	parent.append(container);
	// XXX TODO: Why doesn't container expand to content size in the first place?
	container.css({"width": content.width(),
		"height": content.height()+titlebar.height(),
		"position": "relative"});

	closer = function(ignored) {
		container.hide();
		container.detach();
		content.detach();
		content.hide();
		prevParent.append(content);
		onClose();
	}

	container.find(".closer").bind("click", closer);
	return closer;
}


function output_popupAddSel() {
// pops up the dialog with the additional output items.  The popup
// receives a cleanup function.
	var govButton = output_bussedElements[2];
	var closer = popupInnerWindow(
		$("#genForm-_ADDITEMS"), $("#genForm-_OUTPUT"),
		function() {
			govButton.unbind("click");
			govButton.bind("click", output_popupAddSel);
			govButton.contents()[0].data = "More output fields";
		})
	govButton.unbind("click");
	govButton.bind("click", closer);
	govButton.contents()[0].data = "Pop down field selector";
	return false;
}


function output_itemSelector(pars) {
	// returns a Bussedelement to pop up the element Selector
	var root = $('<button type="button" id="op_additem" class="popButton">More output fields</button>');

	root.bind("click", output_popupAddSel);
	// show nowhere unless there actually are items to show
	var showFor = new Array();
	if (document.getElementById("genForm-_ADDITEMS")) {
		showFor.push('HTML');
	}
	res = output_BussedElement(root, "op_additem", showFor);
	return res;
}


function output_show(el) {
	if (!document.getElementById(el.id)) {
		var dest = $("#genForm-_OUTPUT");
		dest.append(document.createTextNode(" "));
		dest.append(el);
	}
}

function output_hide(el) {
	if (document.getElementById(el.id)) {
		$(el).detach();
	}
}



/////////////////////////////////////////////////////////////////////
// Plotting code -- move (most of this) to something loaded on demand

DATETIME_RE = /^(\d\d\d\d)-(\d\d)-(\d\d)[T ](\d\d):(\d\d):(\d\d.?\d*)$/
DATE_RE = /^(\d\d\d\d)-(\d\d)-(\d\d)$/
function _getValue(s) {
// tries to make some kind of number from a string s
// if s looks like a datetime, return a julian year
// TODO: do something with hours/sexagesimal angles
// if s looks like a float number, return a float
// else return null.
	var dm = DATETIME_RE.exec(s);
	if (dm!=null) {
		var dt = new Date(parseFloat(dm[1]), parseFloat(dm[2]), parseFloat(dm[3]),
			parseFloat(dm[4]), parseFloat(dm[5]), parseFloat(dm[6]));
		return getJYear(dt);
	}
	var dm = DATE_RE.exec(s);
	if (dm!=null) {
		var dt = new Date(parseFloat(dm[1]), parseFloat(dm[2]), parseFloat(dm[3]));
		return getJYear(dt);
	}

	var num = parseFloat(s);
	if (num==num) { // not NaN
		return num;
	}
	return null;
}
	


function _doLinePlot(table, xInd, yInd, style) {
	var data = new Array();
	var options = {"series": {}};

	table.find('tr.data').each(function(index, row) {
		var tds = $(row).children();
		var x = _getValue(tds[xInd].firstChild.data); 
		var y = _getValue(tds[yInd].firstChild.data);
		if (x!=null && y!=null) {
			data.push([x, y]);
		}
	});
	if (style=="Lines") {
		options.series["lines"] = {'show': true};
		data.sort(function(a,b){return a[0]-b[0]});
	} else {
		options.series["points"] = {'show': true};
	}
	jQuery.plot(jQuery('#plotarea'), [data], options);
}


function _makeHistogram(data, numBins) {
	if (data.length<2) {
		return new Array();
	}
	data.sort(function(a,b){return a-b});
	var zp = data[0];
	var binSize = (data[data.length-1]-zp)/numBins;
	if (binSize==0) {
		binSize = 1;
	}

	var histo = new Array();
	for (var i=0; i<numBins; i++) {
		histo.push(0);
	}
	for (index in data) {
		histo[Math.floor((data[index]-zp)/binSize)]++;
	}
	
	data = new Array();
	for (index in histo) {
		data.push([index*binSize+zp, histo[index]]);
	}
	return data;
}

function _doHistogramPlot(table, colInd) {
	var data = new Array();
	table.find('tr.data').each(function(index, row) {
		var val = _getValue($(row).children()[colInd].firstChild.data);
		if (val!=null) {
			data.push(val);
		}
	});

	var histo = _makeHistogram(data, 20);
	jQuery.plot(jQuery('#plotarea'), [{
		bars: {
			show: true,
			barWidth: histo[1][0]-histo[0][0]},
		data: histo}]);
}	


function _doFlotPlot(table, xsel, ysel, usingSel) {
	xInd = xsel.find("option:selected").val()
	yInd = ysel.find("option:selected").val()
	style = usingSel.find("option:selected").val()
	if (yInd=='Histogram') {
		_doHistogramPlot(table, xInd);
	} else {
		_doLinePlot(table, xInd, yInd, style);
	}
}


function _plotUsingFlot(table, options) {
// allows simple plotting of HTML tables.  This only works from
// within openFlotPlot since it uses javascript that's not loaded
// by default.
	// create the plot element
	var plotElement = $('<div id="plotcontainer" style="position:fixed;z-index:3000;background:white;padding-left:3px;padding-right:3px;padding-bottom:3px;border:2px solid gray"><p class="innerTitle"><span class="closer">x&nbsp;</span></p><div id="plotarea" style="width:700px;height:400px;"/></div>');
	plotElement.draggable();
	plotElement.find(".closer").bind("click", function(){
		plotElement.remove()});
	var controlPara = $('<p class="flotControl"></p>');
	plotElement.append(controlPara);

	// Make column selectors from table headings
	var xsel = $('<select/>');
	$(table.find('tr')[0]).find('th').each(function(index, head) {
		xsel.append($('<option value="'+index+'">'+$(head).text()+'</option>'));
	});
	controlPara.append(xsel);
	controlPara.append(' vs. ');
	var ysel = xsel.clone();
	ysel.append($(
		'<option value="Histogram">Histogram</option>'));
	controlPara.append(ysel);
	controlPara.append(' using ');
	var usingSel = $('<select><option selected="selected">Points</option><option>Lines</option></select>')
	controlPara.append(usingSel);

	// Set default plot features from options if given there
	var xselIndex = 0;
	if (options.xselIndex) {
		xselIndex = options.xselIndex;
	}
	var yselIndex = ysel.children().length-1;
	if (options.yselIndex) {
		yselIndex = options.yselIndex;
	}
	var usingIndex = 0;
	if (options.usingIndex) {
		usingIndex = options.usingIndex;
	}
	xsel.children()[xselIndex].setAttribute("selected", "selected")
	ysel.children()[yselIndex].setAttribute("selected", "selected")
	usingSel.children()[usingIndex].setAttribute("selected", "selected")

	// the callback any form items controlling the plot
	var updatePlot = function() {
		try {
			_doFlotPlot(table, xsel, ysel, usingSel);
		} catch (e) {
			jQuery.plot(jQuery('#plotarea'), [{data:[], label: "unplottable"}]);
			throw e;
		}
	}
	xsel.change(updatePlot);
	ysel.change(updatePlot);
	usingSel.change(updatePlot);

	$("body").prepend(plotElement);
	updatePlot();
}

function openFlotPlot(tableElement, options) {
// opens a div that lets you plot some columns of tableElement
	if (options==undefined) {
		options = {};
	}
	$.getScript("/static/js/jquery.flot.js",
		function() {_plotUsingFlot(tableElement, options)});
}


function openVOPlot() {
	votURL = getFormQuery(
		document.getElementById("genForm"), 
			{'_FORMAT': 1, "_TDENC": 1})+
		"&_FORMAT=VOTable&_TDENC=on&_VERB=H";
	window.open(
		'/__system__/run/voplot/fixed?source='+encodeURIComponent(votURL),
		"_self");
}


function insertPreview(node, width) {
// replaces the text content of node with a DC-generated preview
// image.  node has to have a href attribute pointing to a DC
// FITS product for this to work (width is ignored these days).
	if (!node.getAttribute("href")) {
		return;
	}
	var oldHref = node.getAttribute("href");
	var newPars = "preview=True&width="+width;
	var joiner = "?";
	// TODO: do actual URL parsing here
	if (oldHref.indexOf("?")!=-1) { // assume we have a query
		joiner = "&";
	}

	node.removeAttribute("onmouseover");
	$(node).addClass("busy");

	var previewURL = oldHref+joiner+newPars;
	var img = $("<img/>").attr("src", previewURL
		).attr("alt", "[preview image]");

	img.bind("load", function() {
		node.replaceChild(img[0], node.firstChild);
		$(node).removeClass("busy");
	});
	// TODO: use "complete" here when we've updated jquery
	img.bind("error", function() {
		$(node).removeClass("busy");
	});
}

